### PR TITLE
[Clang][Sema] Warn when 'exclude_from_explicit_instantiation' attribute is used on local classes and members thereof

### PR DESCRIPTION
**Pull request by sdkrystian**
A local class and its members declared in a function template are instantiated alongside the definition of that template. It therefore does not make sense to apply the `exclude_from_explicit_instantiation` attribute to such declarations, and this patch adds a warning to diagnose these cases (in addition to ignoring the attribute). 

(The motivation for this patch is to fix a failing test in libc++ for #84050. In particular, [this line](https://github.com/llvm/llvm-project/blob/main/libcxx/include/__memory/uses_allocator_construction.h#L199) in libc++ contains the expression `this->__value_`  (reduced repro [here](https://godbolt.org/z/KqEerKWPd)) which will be looked up prior to instantiation (the lookup context is the current instantiation) once #84050 lands. `_LIBCPP_HIDE_FROM_ABI` includes `__attribute__((exclude_from_explicit_instantiation))`, which TLDR in the reduced repro results in `Local` being instantiated with `Local::operator A` as its `DeclContext`)

### CODE REVIEW

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 1]
```diff
@@ -9315,6 +9330,9 @@ ProcessDeclAttribute(Sema &S, Scope *scope, Decl *D, const ParsedAttr &AL,
   case ParsedAttr::AT_Error:
     handleErrorAttr(S, D, AL);
     break;
+  case ParsedAttr::AT_ExcludeFromExplicitInstantiation:
```
Comment (erichkeane) [inline clang/lib/Sema/SemaDeclAttr.cpp:9333]: What is the previous 'handle' function?  Typically with decl attributes we need have an entry on this switch.

**Review by erichkeane — COMMENTED** [review]
Needs a release note, probably int he 'potentially breaking changes' section?

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 1 above)
Comment (sdkrystian) [reply clang/lib/Sema/SemaDeclAttr.cpp:9333]: `exclude_from_explicit_instantiation` had no 'handle' function prior to this patch -- the only restriction was that it must be applied to the declaration of a variable, function, or class.

Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp:9333]: It still needs to be 'added' to the AST though?  Where was that happening?

Comment (sdkrystian) [reply clang/lib/Sema/SemaDeclAttr.cpp:9333]: In the default case of the switch statement (line 9142, `AL.getInfo().handleDeclAttribute(S, D, AL)`)

**Comment by sdkrystian** [comment]
@erichkeane Perhaps the release note should go under "Attribute Changes in Clang"?

**Comment by erichkeane** [comment]
> @erichkeane Perhaps the release note should go under "Attribute Changes in Clang"?

That would be fine for me as well.

**Comment by sdkrystian** [comment]
Release note added

**Review by erichkeane — APPROVED** [review]

**Comment by mikaelholmen** [comment]
I noticed that when compiling libcxx with this patch there are zillions of warnings like
```
[1950/2073] Building CXX object libcxx/src/CMakeFiles/cxx_shared.dir/filesystem/operations.cpp.o
In file included from /repo/libcxx/src/filesystem/operations.cpp:16:
In file included from /repo/llvm/build-all-builtins/include/c++/v1/iterator:687:
In file included from /repo/llvm/build-all-builtins/include/c++/v1/__iterator/common_iterator.h:31:
/repo/llvm/build-all-builtins/include/c++/v1/variant:912:9: warning: '__exclude_from_explicit_instantiation__' attribute ignored on local class member [-Wignored-attributes]
  912 |         _LIBCPP_HIDE_FROM_ABI void operator()(true_type) const { __this->__emplace<_Ip>(std::forward<_Arg>(__arg)); }
      |         ^
/repo/llvm/build-all-builtins/include/c++/v1/__config:792:22: note: expanded from macro '_LIBCPP_HIDE_FROM_ABI'
  792 |       _LIBCPP_HIDDEN _LIBCPP_EXCLUDE_FROM_EXPLICIT_INSTANTIATION                                                       \
      |                      ^
/repo/llvm/build-all-builtins/include/c++/v1/__config:695:72: note: expanded from macro '_LIBCPP_EXCLUDE_FROM_EXPLICIT_INSTANTIATION'
  695 | #    define _LIBCPP_EXCLUDE_FROM_EXPLICIT_INSTANTIATION __attribute__((__exclude_from_explicit_instantiation__))
      |                                                                        ^
```
Is there any plan to clean this out?

**Comment by sdkrystian** [comment]
@mikaelholmen I believe there are two uses of `exclude_from_explicit_instantiation` in libc++ which are responsible for all these warnings. I was assuming that someone more experienced with the libc++ side of things would handle it, but I can submit a patch that addresses these myself.

**Comment by mikaelholmen** [comment]
Nice! I have no idea about this I just saw the warnings when I compiled and connected them to this patch.

**Pull request merged**